# Logistic Regression on Heart Disease prediction

#### Jun 10 2019

In [25]:
# Loading the dependencies for the dataset
import numpy as np
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt

In [26]:
# Loading the dataset
heart_data = pd.read_csv("E:/Data Scientist/Datasets/framingham_heart_disease.csv",sep=",")
print(heart_data.head())
print(heart_data.shape)

   male  age  education  currentSmoker  cigsPerDay  BPMeds  prevalentStroke  \
0     1   39        4.0              0         0.0     0.0                0   
1     0   46        2.0              0         0.0     0.0                0   
2     1   48        1.0              1        20.0     0.0                0   
3     0   61        3.0              1        30.0     0.0                0   
4     0   46        3.0              1        23.0     0.0                0   

   prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  heartRate  glucose  \
0             0         0    195.0  106.0   70.0  26.97       80.0     77.0   
1             0         0    250.0  121.0   81.0  28.73       95.0     76.0   
2             0         0    245.0  127.5   80.0  25.34       75.0     70.0   
3             1         0    225.0  150.0   95.0  28.58       65.0    103.0   
4             0         0    285.0  130.0   84.0  23.10       85.0     85.0   

   TenYearCHD  
0           0  
1           0  
2 

In [27]:
# From the dataset education feature is not related to the model analysis hence drop the feature
heart_data = heart_data.drop(['education'], axis = 1, inplace= False)

In [28]:
heart_data.head()

male  age  currentSmoker  cigsPerDay  BPMeds  prevalentStroke  \
0     1   39              0         0.0     0.0                0   
1     0   46              0         0.0     0.0                0   
2     1   48              1        20.0     0.0                0   
3     0   61              1        30.0     0.0                0   
4     0   46              1        23.0     0.0                0   

   prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  heartRate  glucose  \
0             0         0    195.0  106.0   70.0  26.97       80.0     77.0   
1             0         0    250.0  121.0   81.0  28.73       95.0     76.0   
2             0         0    245.0  127.5   80.0  25.34       75.0     70.0   
3             1         0    225.0  150.0   95.0  28.58       65.0    103.0   
4             0         0    285.0  130.0   84.0  23.10       85.0     85.0   

   TenYearCHD  
0           0  
1           0  
2           0  
3           1  
4           0

In [29]:
# Check for the initial statstics analysis
heart_data.describe()

male          age  currentSmoker   cigsPerDay       BPMeds  \
count  4238.000000  4238.000000    4238.000000  4209.000000  4185.000000   
mean      0.429212    49.584946       0.494101     9.003089     0.029630   
std       0.495022     8.572160       0.500024    11.920094     0.169584   
min       0.000000    32.000000       0.000000     0.000000     0.000000   
25%       0.000000    42.000000       0.000000     0.000000     0.000000   
50%       0.000000    49.000000       0.000000     0.000000     0.000000   
75%       1.000000    56.000000       1.000000    20.000000     0.000000   
max       1.000000    70.000000       1.000000    70.000000     1.000000   

       prevalentStroke  prevalentHyp     diabetes      totChol        sysBP  \
count      4238.000000   4238.000000  4238.000000  4188.000000  4238.000000   
mean          0.005899      0.310524     0.025720   236.721585   132.352407   
std           0.076587      0.462763     0.158316    44.590334    22.038097   
min           0.000000      0.000000     0.000000   107.000000    83.500000   
25%           0.000000      0.000000     0.000000   206.000000   117.000000   
50%           0.000000      0.000000     0.000000   234.000000   128.000000   
75%           0.000000      1.000000     0.000000   263.000000   144.000000   
max           1.000000      1.000000     1.000000   696.000000   295.000000   

             diaBP          BMI    heartRate      glucose   TenYearCHD  
count  4238.000000  4219.000000  4237.000000  3850.000000  4238.000000  
mean     82.893464    25.802008    75.878924    81.966753     0.151958  
std      11.910850     4.080111    12.026596    23.959998     0.359023  
min      48.000000    15.540000    44.000000    40.000000     0.000000  
25%      75.000000    23.070000    68.000000    71.000000     0.000000  
50%      82.000000    25.400000    75.000000    78.000000     0.000000  
75%      89.875000    28.040000    83.000000    87.000000     0.000000  
max     142.500000    56.800000   143.000000   394.000000     1.000000

 From first statstics if we consider feature "cigsPerDay" the max value is 70 and min value is 0 but the mean value is 9.003089 which is not even close to the 50% data. 

In [30]:
# Look for the missing values in the dataset
heart_data.isnull().sum()

male                 0
age                  0
currentSmoker        0
cigsPerDay          29
BPMeds              53
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol             50
sysBP                0
diaBP                0
BMI                 19
heartRate            1
glucose            388
TenYearCHD           0
dtype: int64

In [31]:
# Let check the percent of missing value for the dataset 
mis_count = 0
for i in heart_data.isnull().sum(axis=1):
    if i>0:
        mis_count = mis_count+1
print("total no of missing value in the dataset :", mis_count)
print("Percent of missing data in dataset when compared to whole dataset", round(mis_count/(len(heart_data.index))*100))

total no of missing value in the dataset : 489
Percent of missing data in dataset when compared to whole dataset 12


From the above cell missing value contribute 12% of the whole dataset.
Hence we can neglect the missing data and load the dataset without misssing data

In [32]:
# Dropping the 'NA' values in the dataset
heart_data.dropna(axis=0,inplace=True)

In [33]:
# Replacing the missing values with mean for respective column
#for col in heart_data.columns:
    #heart_data = heart_data.fillna(heart_data1[col].mean())
#print(heart_data.isnull().sum())
#heart_data1.head()
# Convert any nan or inf values to the float  
#heart_data1=heart_data1[~heart_data1.isin([np.nan,np.inf,-np.inf]).any(1)].astype(np.float64)
cols = heart_data.columns[:-1]
heart_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3749 entries, 0 to 4237
Data columns (total 15 columns):
male               3749 non-null int64
age                3749 non-null int64
currentSmoker      3749 non-null int64
cigsPerDay         3749 non-null float64
BPMeds             3749 non-null float64
prevalentStroke    3749 non-null int64
prevalentHyp       3749 non-null int64
diabetes           3749 non-null int64
totChol            3749 non-null float64
sysBP              3749 non-null float64
diaBP              3749 non-null float64
BMI                3749 non-null float64
heartRate          3749 non-null float64
glucose            3749 non-null float64
TenYearCHD         3749 non-null int64
dtypes: float64(8), int64(7)
memory usage: 468.6 KB


In [34]:
#from statsmodels.tools import add_constant as add_constant
# This add a column particularly for the intercept
#heart_data_Wconst = add_constant(heart_data)
#heart_data_Wconst.head()

In [35]:
# Now build the logistic model
logit_model = sm.Logit(heart_data.TenYearCHD,heart_data[cols]).fit()
logit_model.summary()

Optimization terminated successfully.
         Current function value: 0.400181
         Iterations 6


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:             TenYearCHD   No. Observations:                 3749
Model:                          Logit   Df Residuals:                     3735
Method:                           MLE   Df Model:                           13
Date:                Tue, 11 Jun 2019   Pseudo R-squ.:                 0.06313
Time:                        19:57:16   Log-Likelihood:                -1500.3
converged:                       True   LL-Null:                       -1601.4
                                        LLR p-value:                 4.824e-36
===================================================================================
                      coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
male                0.4093      0.103      3.980      0.000       0.208       0.611
age                 0.0285      0.006      4.977      0.000       0.017       0.040
currentSmoker      -0.2377      0.150     -1.581      0.114      -0.532       0.057
cigsPerDay          0.0214      0.006      3.530      0.000       0.010       0.033
BPMeds              0.3233      0.227      1.422      0.155      -0.122       0.769
prevalentStroke     0.7832      0.483      1.621      0.105      -0.164       1.730
prevalentHyp        0.9720      0.122      7.967      0.000       0.733       1.211
diabetes            0.7821      0.291      2.685      0.007       0.211       1.353
totChol            -0.0015      0.001     -1.394      0.163      -0.004       0.001
sysBP               0.0133      0.004      3.544      0.000       0.006       0.021
diaBP              -0.0300      0.006     -5.009      0.000      -0.042      -0.018
BMI                -0.0466      0.012     -3.809      0.000      -0.071      -0.023
heartRate          -0.0228      0.004     -5.814      0.000      -0.030      -0.015
glucose             0.0019      0.002      0.900      0.368      -0.002       0.006
===================================================================================
"""

Over all model gives the p-value as 0 which is less than alpha value 0.05 or 5%. Now we go with back elimination to check the respective logistic regression and combinations of each feature.

In [22]:
def back_elimination_feature(dataset,output_var,col_list):
    while len(col_list)>0:
        model = sm.Logit(output_var,dataset[col_list]).fit(disp=0)
        result = model
        large_pvalue = round(result.pvalues,3).nlargest(1)
        if large_pvalue[0]<0.05:
            return result
            break
        else:
            col_list = col_list.drop(large_pvalue.index)

In [23]:
result = back_elimination_feature(heart_data,heart_data['TenYearCHD'],cols)

In [24]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:             TenYearCHD   No. Observations:                 3749
Model:                          Logit   Df Residuals:                     3740
Method:                           MLE   Df Model:                            8
Date:                Tue, 11 Jun 2019   Pseudo R-squ.:                 0.05994
Time:                        19:56:15   Log-Likelihood:                -1505.4
converged:                       True   LL-Null:                       -1601.4
                                        LLR p-value:                 3.129e-37
================================================================================
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
male             0.4169      0.102      4.093      0.000       0.217       0.617
age              0.0269      0.005      4.911      0.000       0.016       0.038
cigsPerDay       0.0137      0.004      3.366      0.001       0.006       0.022
prevalentHyp     1.0225      0.118      8.657      0.000       0.791       1.254
diabetes         0.9639      0.227      4.249      0.000       0.519       1.408
sysBP            0.0137      0.004      3.684      0.000       0.006       0.021
diaBP           -0.0311      0.006     -5.230      0.000      -0.043      -0.019
BMI             -0.0467      0.012     -3.876      0.000      -0.070      -0.023
heartRate       -0.0245      0.004     -6.504      0.000      -0.032      -0.017
================================================================================
"""

As we see, these are features which can be helpful to find the logistic regression of this heart data but "totchol" was eliminated in this but we have the influencing record which cause the feature to eliminate because its p-value is > 0.05. If we remove the influencing record from this feature it can be useful for logistic regression which will be done in next part but right now taking the all the above the features with totchol we do logistic regression using sklearn library 

In [52]:
# lets separate the input features and target into separate 
feature_taken = heart_data[['male','age','cigsPerDay','prevalentHyp','diabetes','sysBP','diaBP','BMI','heartRate','totChol']]
X=feature_taken
y=heart_data[['TenYearCHD']]

In [53]:
# Lets split these into train and test set
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=52)
print("X Train set :",X_train.shape[0])
print("y Train set:", y_train.shape[0])
print("X Test set:", X_test.shape[0])
print("y Test set:", y_test.shape[0])

X Train set : 2999
y Train set: 2999
X Test set: 750
y Test set: 750


In [55]:
# Load the linear model from sklearn
from sklearn.linear_model import LogisticRegression
# insantiate the Logistic Regression
model = LogisticRegression()
model.fit(X_train,Y_train)
y_pred = model.predict(X_test)

c:\users\gudis\appdata\local\programs\python\python36\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
c:\users\gudis\appdata\local\programs\python\python36\lib\site-packages\sklearn\utils\validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [56]:
# Now we check for the accuracy of the model
from sklearn import metrics
Confusion_Matrix = metrics.confusion_matrix(y_pred,y_test)
Confusion_Matrix

array([[627, 113],
       [  2,   8]], dtype=int64)

In [58]:
# Now the accuracy of the model 
accuracy = metrics.accuracy_score(y_test,y_pred)
print("Accuracy of the model:",accuracy)

Accuracy of the model: 0.8466666666666667
